### DISCUSSION TITLE
Rethink of Ligature 再思连字

### DISCUSSION THREAD

**Discussion by subframe7536** [subframe7536/maple-font#240] [Ideas]
> [!NOTE]
> All the content in the article is based on my personal non-professional practical experience, and there may be inaccuracies. Please feel free to leave comments and corrections.

[中文版本](#discussioncomment-10512385)

## Background

Ligature is an OpenType feature that allows for context-based rules to replace characters and enables users to selectively enable or disable certain features (such as `cv01` or `ss01`). It possesses powerful capabilities that can greatly enhance the reading experience of code. Ligatures were also one of the main driving forces behind the creation of this font.

### Why Create This Font

The first ligature font I encountered was the well-known `Fira Code`. At that time, I had just started learning programming and had been using the default `Consolas` font. The discovery of `Fira Code` added a touch of novelty to my dry learning process of the C language. Seeing the final code after `AC` gave me a sense of accomplishment, especially when I noticed that people who is failed to solve the problem around me were still using sans-serif fonts. However, I have to admit that as a beginner, using fonts with significant changes like `==` / `!=` posed a bit of a learning curve, often causing confusion.

Later on, I came across other ligature fonts such as `Jetbrains Mono`, `Iosevka`, `Victor Mono`, and `Cascadia Code`. My perspective continued to broaden, but at the same time, I was not completely satisfied with some of the glyph shapes. This dissatisfaction led me to the idea of creating my own font.

### Versions 5 and 6

In `Maple Mono` versions 5 and 6, the ligature rules were modified by me based on the `fea` script code parsed from `Fira Code Retina` in Font Creator. At that time, as I was just starting out with font development, I could only imitate the process. In the end, I managed to cobble together a font that was passable, and I posted it on v2ex.com . To my surprise and gratitude, it received enthusiastic support and feedback from the community, gaining over 1,000 stars in three days. Encouraged by this response, I quickly addressed the reported issues by releasing a fixed version and subsequently several optimized versions.

## Current Situation and Thoughts

In `Maple Mono` version 7, I made modifications to the `Jetbrains Mono` variable font source files to suit my personal preferences. I utilized the powerful font tool library @ftcli for optimization. In terms of ligatures, a significant portion was directly reused with adjustments made for many edge cases. Due to a disruptive standardization of the ligature rules in version 6.2, I took a more conservative approach in the current version 7 Beta, sticking to the previous method: using all ligatures that align with my personal preferences by default, and using feature toggles to disable glyphs that may confuse others.

As a result, a new question emerged: how can I provide a **more customizable ligature solution** that allows others to compile their own fonts without directly modifying them, through configuration options? One idea I have is: **using common glyphs as defaults and making all personalized glyphs available as toggle options for users to enable or disable, or provide a fixed toggle setting through scripts/websites**.

### Selection of Ligature Features

Ligature features can be in the form of Character Variants (`cv01` - `cv99`) or Stylistic Sets (`ss01` - `ss20`), totaling 119. Functionally, there is no difference between the two, only a distinction in naming conventions for different types of features. However, for the sake of backward compatibility, fonts like `Fira Code` and `Jetbrains Mono` did not adhere to this convention. Currently, `Maple Mono` 's ligatures also do not entirely comply with this convention, so reconsideration will be needed when restructuring the feature toggles.

### Compile in Browser

Due to the browser's excellent cross-platform compatibility and convenience, more and more people prefer to complete tasks within the browser. Last year, I discovered [`Commit Mono`](https://github.com/eigilnikolajsen/commit-mono), which allows for browser-based font compilation using the `opentype.js` library. However, after developing and compiling with this library, I found that the resulting file size exceeded that generated using `fonttools` in `Python` by more than twice. Due to my limited resources, I chose not to delve deeper into this area after trying other popular JavaScript libraries like `fonteditor-core` and `fontkit`. If you are keen on compiling in the browser, I recommend using [fontfreeze](https://mutsuntsai.github.io/fontfreeze/), a tool that uses `Pyodide` to convert `Python` code to `WASM` for frontend use. However, please note that it cannot compile `Maple Mono NF`.

**Comment by subframe7536** [comment]
> [!NOTE]
> 文章中所有的内容都是基于我个人非专业的实践经验，可能存在不准确的地方，欢迎留言并指正
## 背景

连字是一个 OpenType 特性，它可以制定基于上下文的规则来对字符进行替换，还可以让用户自行启用或关闭部分特性（如 `cv01`、`ss01` 等），具备十分强大的能力，能够较大程度地提升代码阅读的体验，连字也是我制作这个字体的主要动力之一。

### 为什么要做这个字体

我最开始接触到的连字字体，是著名的 `Fira Code`，彼时我刚刚接触编程，一直用着默认的 `Consolas`，`Fira Code` 的发现，让我枯燥的 C 语言学习过程中多了几分新鲜，看着 AC 后的代码觉得很有成就感，特别是看到周围做不出题的人还在用着无衬线的字体的时候。但是我不得不承认，在新手阶段，使用类似 `==` /  `!=` 这种变化较大的字体时，还是有一些门槛的，经常会认不出来。

再后来，我接触到了其他的连字字体，比如 `Jetbrains Mono`、`Iosevka`、`Victor Mono`，`Cascadia Code`，我的思路不断被打开，但同时也对其中一部分的字形不太满意，于是萌发了自己做字体的想法。

### V5 和 V6

在 `Maple Mono` V5 和 V6 中，连字的规则是我根据 `Fira Code Retina` 在 `Font Creator` 中解析后的 `fea` 脚本代码修改而来的，当时刚刚上手字体开发，能做的也只是照猫画虎，最终东拼西凑出来一个字体，用着也还行，就发到了 v2ex.com 上，结果受到大家的热烈支持和反馈，三天涨了 1k 多 star。我深受其鼓舞，也是很感谢大家的支持，于是很快针对提出的问题发布了一个修复的版本，并在后续发布了几个优化的版本。

## 现状和思考

在 `Maple Mono` V7 中，我基于 `Jetbrains Mono` 可变字体的源文件，进行了符合我个人喜好的修改，并使用 @ftcli 这个强大的字体工具库进行优化。在连字方面，有很大一部分是直接复用的，对很多边界情况做了处理。由于之前 V6.2 的时候连字规则进行过一次破坏性的规范化处理，所以在目前的 V7 Beta 中就相对保守，直接沿用了之前的思路：**在默认的情况下使用所有符合我个人偏好的连字，使用特性开关来关闭这些可能造成他人困惑的字形**。

于是，自然而然地，一个问题出现了：如何提供一个**自定义能力更强**的连字方案，让其他人无需直接修改字体，通过配置的方式自行编译一个属于自己的字体？目前我的一个想法：**将通用的字形作为默认，将所有的个性化字形作为开关选项，供使用者启用关闭，或者通过脚本/网站编译，设置为固定开启**

### 特性开关的选择

特性开关可以是 Character Variants（`cv01`-`cv99`），也可以是 Stylistic Sets（`ss01`-`ss20`），总共 119 个，两者在功能实现上没有任何区别，只是在约定上用于区分不同类型的特性开关。但是，由于向后兼容性的需要，`Fira Code`、`Jetbrains Mono` 等字体并没有遵守这个约定。在目前 `Maple Mono` 的连字中，也并没有完全遵守这个约定，因此在重新构建特性开关时也需要考虑。

### 浏览器中构建

由于浏览器良好的跨平台兼容性和便携性，越来越多人希望在浏览器中完成一切事情。在去年的时候，我发现了 [`Commit Mono`](https://github.com/eigilnikolajsen/commit-mono) 实现了在浏览器中构建，使用的是 `opentype.js` 这个库。但是当我在使用这个库开发并构建后发现，构建后的体积超出了使用 `python` 的 `fonttools` 构建的体积的一倍多，由于我个人精力有限，在尝试 `fonteditor-core`、`fontkit` 等其他流行程度高的 js 库之后，我不打算在这其中深入研究。如果你们很想在浏览器中构建，推荐使用 [fontfreeze](https://mutsuntsai.github.io/fontfreeze/) 这个工具，它使用了 `Pyodide` 将 `Python` 代码转化为 `WASM` 供前端调用。但是需要注意，`Maple Mono NF` 是无法构建的。

**Comment by rocascent** [comment]
连字作为一种上下文无关的特性，在代码中确实容易出现不协调的情况。

**Reply by rocascent** [reply]
再考虑到不同编程语言中符号又有不同的词法或语义，感觉很难做得完美。

**Reply by subframe7536** [reply]
是的，特别是在正则表达式里，简直头大

**Comment by KkemChen** [comment]
话说 连体字 什么时候打算支持 比如 ==  != 之类的🤗

**Reply by subframe7536** [reply]
一直支持的啊

**Reply by KkemChen** [reply]
啊 怎么开启

**Reply by subframe7536** [reply]
v6要通过font feature开启，v7提供默认的版本
